### PR TITLE
[8.0] feat: initial IAM integration tests

### LINKED ISSUE #6625: Expand integration tests for IaM
The integration tests (https://github.com/DIRACGrid/DIRAC/actions/workflows/integration.yml) should be expanded with interactions with IAM and CheckIn. 

For IAM, a docker container is provided: https://indigo-iam.github.io/v/v1.8.0/docs/getting-started/docker/
I could not find one for CheckIn yet.

This starts by adding to https://github.com/DIRACGrid/DIRAC/blob/integration/tests/CI/docker-compose.yml

### PR DESCRIPTION
**Pull request by aldbr**
This PR aims at adding unit & integration tests around `Resources/IdProviders` components.
Integration tests interact with an IAM docker container.

Related to https://github.com/DIRACGrid/DIRAC/issues/6625

BEGINRELEASENOTES
*Resources
NEW: unit tests for IdProviders
*tests
NEW: integration tests for IdProviders
ENDRELEASENOTES

### CODE REVIEW

File: src/DIRAC/Resources/IdProvider/Utilities.py [hunk 1]
```diff
@@ -3,78 +3,62 @@
...
-    for section in result["Value"]:
-        if section.endswith("Providers"):
-            result = getProvidersForInstance(section[:-9])
-            if not result["OK"]:
-                return result
-            if provider in result["Value"]:
-                return gConfig.getOptionsDictRecursively(f"/Resources/{section}/{provider}/")
-    return S_ERROR(f"{provider} provider not found.")
+
+    return S_OK(result["Value"])
```
Comment (fstagni) [inline src/DIRAC/Resources/IdProvider/Utilities.py]: I think the block from line 60 is equivalent to do

```
return gConfig.getOptionsDict(f"/Resources/IdProviders/{idProvider}")
```

File: .github/workflows/integration.yml [hunk 2]
```diff
@@ -68,6 +68,8 @@ jobs:
       run: ./integration_tests.py test-client || touch client-tests-failed
     - name: Elasticsearch logs
       run: docker logs elasticsearch
+    - name: IAM logs
+      run: docker logs iam-login-service
```
Comment (fstagni) [inline .github/workflows/integration.yml]: For my info, why is this needed?

File: src/DIRAC/Resources/IdProvider/Utilities.py (hunk 1 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/Utilities.py]: Indeed, I am going to fix this.
Do we even need this function then? (`getIdProviderParameters`)

File: .github/workflows/integration.yml (hunk 2 above)
Comment (aldbr) [reply .github/workflows/integration.yml]: Well, it is not essential actually.
Should I remove these lines?

Comment (aldbr) [reply .github/workflows/integration.yml]: I removed these lines.

File: src/DIRAC/Resources/IdProvider/CheckInIdProvider.py [hunk 3]
```diff
@@ -16,8 +16,13 @@ def getGroupScopes(self, group):
...
-                getVOForGroup(group),
-                group.split("_")[1],
-            )
+            vo = getVOForGroup(group)
+            if not vo:
+                return []
+
+            groupElements = group.split("_")
+            if len(groupElements) < 2:
+                return []
```
Comment (fstagni) [inline src/DIRAC/Resources/IdProvider/CheckInIdProvider.py:25]: The "DIRAC group" is something like "lhcb_user" or "dteam_prod" I guess...? But there's no rule (that I know) where this convention is enforced, or I am wrong? (this has of course nothing to do with this PR, just I happen to notice this only now)
In LHCb production we (historically) have for example a "diracAdmin" group.

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py [hunk 4]
```diff
@@ -16,5 +16,13 @@ def getGroupScopes(self, group):
...
-            idPScope = f"wlcg.groups:/{getVOForGroup(group)}/{group.split('_')[1]}"
+            vo = getVOForGroup(group)
+            if not vo:
+                return []
+
+            groupElements = group.split("_")
+            if len(groupElements) < 2:
+                return []
+
+            idPScope = f"wlcg.groups:/{vo}/{groupElements[1]}"
```
Comment (fstagni) [inline src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Again, nothing to do with your PR, but this `"wlcg.groups"` seems just an hardcoded value...?

**Review by fstagni — COMMENTED** [review]
You made only one commit for effectively many changes/fixes, for next time you can probably split them into different commits.

File: src/DIRAC/Resources/IdProvider/Utilities.py (hunk 1 above)
Comment (fstagni) [reply src/DIRAC/Resources/IdProvider/Utilities.py]: It might not be needed indeed.

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Yes indeed.
The `wlcg.groups` scope is actually bound to the WLCG token profile that IAM supports.
But while I was playing with the IAM configuration, I noticed that it was not the only supported profile.
See here: https://indigo-iam.github.io/v/v1.8.0/docs/reference/configuration/jwt-profiles/
So the name should actually be something like: `WLCGIdProvider` or `IAMWLCGIdProvider`.
I think for now it is okay but we should keep that information in mind.

I made a small list of questions like that while working on these tests.
I should create some issues related to them, or may be create a small presentation to suggest some changes in the code.

File: src/DIRAC/Resources/IdProvider/CheckInIdProvider.py (hunk 3 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/CheckInIdProvider.py:25]: Indeed, I have no idea whether there is a rule for that.

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: I realized that I did not really answer the question.

`wlcg.groups` is specific to the WLCG token [profile](https://github.com/WLCG-AuthZ-WG/common-jwt-profile/blob/master/profile.md#scope-based-group-selection) as I said.
Users can define it in the scope in their token request:

```bash
$ curl ... -d scope=wlcg.groups
$ curl ... -d scope=wlcg.groups:/dirac/user
```

And they will receive an access token embedding a `wlcg.groups` claim defining the groups:

```
{
  "sub": ...
  "iss": ...
  "wlcg.groups":
  [
    "dirac",
    "dirac/user" 
  ] 
}
``` 

This claim might be checked by external resources before accepting user requests.
Therefore, `wlcg.groups` should be added to the scope.
And because this can be different for each token profile, we have one class for each one.

Is it clear enough?

Comment (fstagni) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Then maybe rename the class to something more meaningful?

Does it mean that other VOs might need to employ a different profile?

Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Since the discovery of the endpoints is well defined and standardized (https://openid.net/specs/openid-connect-discovery-1_0.html), there should not be any problem to retrieve the endpoints to interact with different IdProviders (meaning we would use `OAuth2IdProvider` to interact with `Indigo IAM`, `CheckIn`, ...). The main difference is the token profile used.

`VO_a` could theoretically use `Indigo IAM` with the `WLCG` token profile, while `VO_b` could use `Indigo IAM` with the `AARC` token profile and `VO_c` would use `CheckIn` with the `AARC` token profile.
Thus, I propose the following structure:

```
- Resources
  - IdPClient (currently IdProvider)
    - OAuth2IdPClient (currently OAuth2IdProvider - would embed a TokenProfile object)
    - TokenProfile
      - TokenProfile (abstract class defining getGroupScope() and getScopeGroup())
      - WLCGTokenProfile (currently IAMIdProvider)
      - AARCTokenProfile (currently CheckInProvider)
```

What do you think?
@atsareg

File: src/DIRAC/Resources/IdProvider/CheckInIdProvider.py (hunk 3 above)
Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/CheckInIdProvider.py:25]: There is no rule for that. But this can be the moment to introduce it. Would be nice to see the cost of passing diracAdmin -> dirac_admin

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: I agree in general with the following remarks:
1. I do not think that IdProvider should change to IdPClient. All the DIRAC resources are actually clients of a particular third party service, however we name corresponding classes by rather the service nature related names. So, let's be consistent.
2. TokenProfile should be then made as loadable through ObjectLoader to be possible to develop these objects in extensions.

Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: 1. I agree that most DIRAC resources are actually clients of a particular third party service. But I think this case is a bit different:
- Generally `IdProvider` refers to an `IAM` or `CheckIn` instance. But multiple OAuth2 clients can be defined and used to interact with a same "IdProvider" (`IAM` or `CheckIn`).
- `Client` is actually a standard OAuth2 term.

2. Yes sure.

File: src/DIRAC/Resources/IdProvider/IdProviderFactory.py [hunk 5]
```diff
@@ -1,19 +1,14 @@
...
+from DIRAC import S_OK, S_ERROR, gLogger, gConfig
 from DIRAC.Core.Utilities import ObjectLoader, ThreadSafe
-from DIRAC.Core.Utilities.DictCache import DictCache
-from DIRAC.Resources.IdProvider.Utilities import getProviderInfo, getSettingsNamesForIdPIssuer
 from DIRAC.FrameworkSystem.private.authorization.utils.Clients import getDIRACClients
 from DIRAC.FrameworkSystem.private.authorization.utils.Utilities import collectMetadata
+from DIRAC.Resources.IdProvider.Utilities import getIdProviderIdentifierFromIssuerAndClientID
+
 
 gCacheMetadata = ThreadSafe.Synchronizer()
```
Comment (atsareg) [inline src/DIRAC/Resources/IdProvider/IdProviderFactory.py]: If you remove the cacheMetadata which is not currently used then you should remove this line as well together with the corresponding import

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: I think as with other resources we need here hierarchical description with corresponding objects in the code. Then the configuration should have rather the following structure:

     Resources
     {
        IdProviders
        {
            CheckIn
            {
                issuer = ...
                ...
                Clients
                {
                    Biomed
                    {
                        ClientID = ...
                        ClientSecret = ...
                        TokenHandler = BiomedTokenHandler
                        VO = biomed 
                    }
                    Galaxy
                    {
                        ClientID = ...
                        ClientSecret = ...
                        TokenHandler = DteamTokenHandler
                        VO = usegalaxy 
                    }
                } 
            } 
            WLCG_IAM
            {
                issuer = ...
                Clients
                {
                    DTeam
                    {
                        ...
                    }
                } 
            }
        }
      }

So, IdProvider stays and TokenProfile becomes IdPClient which has special methods for managing token profiles. The IdPClients can be configured to be used by limited number of VOs. The VO description in the Registry can have reference to the IdProvider as for example:

    Registry
    {
       VO
       {
          biomed
          {
             IdProvider = Check_In:::Biomed
          }
       }
    }

Note also that we should have a generic IdPClient class where group<->scopes mapping rules are simply defined in the configuration as one-to-one pairs. This should be better defined in the Registry, e.g.

    Registry
    {
      VO
      {
         biomed
         {
            IdProvider = Check_In
            TokenProfile
            {
                biomed_user = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=member#aai.egi.eu
                biomed_pilot = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=pilot#aai.egi.eu
            } 
         }
      }
    }

I think this will capture (almost) all the cases and no specialized TokenProfile classes will be needed at all. We can leave only with standard OAuth2IdProvider class configurable for all the VO needs.

File: src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py [hunk 6]
```diff
@@ -224,70 +366,57 @@ def getScopeGroups(self, scope: str) -> list[str]:
...
-        if not refresh_token:
-            refresh_token = self.token.get("refresh_token")
         try:
-            token = self.refresh_token(self.get_metadata("token_endpoint"), refresh_token=refresh_token, **kwargs)
+            token = self.refresh_token(self.get_metadata("token_endpoint"), refresh_token=refreshToken, **kwargs)
             return S_OK(OAuth2Token(dict(token)))
-        except Exception as e:
-            self.log.exception(e)
-            return S_ERROR(repr(e))
+        except OAuthError as e:
```
Comment (atsareg) [inline src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py:418]: Are you sure there can not be other exceptions than this specific one ? They will pass uncaught if there is no general exception catching

File: src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py [hunk 7]
```diff
@@ -304,128 +433,55 @@ def fetchToken(self, **kwargs):
...
                 subject_token_type="urn:ietf:params:oauth:token-type:access_token",
                 scope=scope,
             )
-            if not token:
-                return S_ERROR(f"Cannot exchange token with {scope} scope.")
             return S_OK(OAuth2Token(dict(token)))
-        except Exception as e:
-            self.log.exception(e)
-            return S_ERROR(f"Cannot exchange token with {scope} scope: {repr(e)}")
+        except OAuthError as e:
```
Comment (atsareg) [inline src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py:466]: Similar comment as above, also for some other places

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Well, the scope<->group definition is now in the Registry/Group/{group}/IdPRole . This was meant for defining extra scopes, e.g. compute scopes. But this can contain everything necessary for  scope<->group mapping. Should be better renamed to IdPMapping or IdPScope. Then the mapping will be done only in Group definition rather than in the VO section. BTW, this will allow diracAdmin group to survive.

File: src/DIRAC/Resources/IdProvider/CheckInIdProvider.py (hunk 3 above)
Comment (chaen) [reply src/DIRAC/Resources/IdProvider/CheckInIdProvider.py:25]: There surely shouldn't be any rule like that, that is way too constraining.

Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/CheckInIdProvider.py:25]: Agreed. We should rather make explicit mapping of dirac group names to IdP entitlements in the group configuration

**Comment by atsareg** [comment]
I think we can drop specific IAMIdProvider and CheckInIdProvider now and make no assumptions on how claims/scopes should be parsed, just define them explicitly in the configuration. If later there will be some real differences in the logic of different IdPs then this can be reintroduced. This will also simplify claimParser() utility (if it will be at all needed) and parseEduperson() method.

File: src/DIRAC/Resources/IdProvider/IAMIdProvider.py (hunk 4 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: Example1: 

> I think as with other resources we need here hierarchical description with corresponding objects in the code. Then the configuration should have rather the following structure

Ok I agree with the structure, it seems better.
Just a few remarks to make sure we are on the same page (correct me if I don't understand correctly): 
- An `IdProviders` subsection (`CheckIn` or `WLCG_IAM` in your first example) corresponds to an instance of an IdP, which could handle, IIUC, one or many VOs.
- In your first example, you defined one client per VO (`Biomed`, `Galaxy`), but we could actually have multiple clients for a same VO: `Biomed_Pilot` the client used to submit pilots and `Biomed_Admin` the client used to perform token exchange for instance. 
- In your first example, the `TokenHandler` option corresponds to the token profile.

Example2:
> So, IdProvider stays and TokenProfile becomes IdPClient which has special methods for managing token profiles. The IdPClients can be configured to be used by limited number of VOs. The VO description in the Registry can have reference to the IdProvider as for example

Not sure to understand why `TokenProfile` becomes `IdPClient`? What is `IdPClient`? 

Does your second example mean that we have one client per VO? (+ another one defined in `Operations/<VO>/Pilot`?)
Because we are dealing with a OAuth2 client, may be the CS option should be named `IdProviderClient` instead of `IdProvider`, what do you think? 

Example3:

> Note also that we should have a generic IdPClient class where group<->scopes mapping rules are simply defined in the configuration as one-to-one pairs. This should be better defined in the Registry
> I think this will capture (almost) all the cases and no specialized TokenProfile classes will be needed at all. We can leave only with standard OAuth2IdProvider class configurable for all the VO needs.

Does it mean that we would have no `TokenProfile` class? Would the specificity of a token profile be entirely defined in the configuration then?
It seems like the `TokenProfile` section in your example correspond to group-based scopes. What if a VO does not support group-based scope? They would then define something like `biomed_pilot = compute.read compute.write...`. Therefore, I think that the section should actually be named `GroupScope`.

But the question is: isn't it the purpose of the `IdPRole` option (defined in `Groups/<VO>/IdPRole`)?

Except from that, if we have one token profile per VO, then may be it is better to define the token profile specific scopes in the configuration.
But I am not sure of that: for instance in EGI, VOs and `CheckIn` are using the `AARC` token profile, am I wrong?
LHC experiments use the `WLCG` token profile.
If multiple VOs use the same token profile, I think it is easier for them to add the name of a token profile rather than defining manually the token profile specific scopes in the CS (the `eduperson_entitlement` scope is rather complex to write).

Here is a suggestion combining my remarks:

```
Registry
{
  VO
  {
    biomed
    {
      IdProviderClient = Check_In:::Biomed_User
    }
    galaxy
    {
      IdProviderClient = Check_In:::Galaxy_User
    }  
  }
  Groups
  {
    biomed_user
    {
      # this would be parsed by a AARCTokeProfile instance
      GroupScope = biomed_member
      # if there is one token profile per VO, which does not seem to be the case, here is an alternative
      # IdPScope = eduperson_entitlement?value=urn:mace:egi.eu:group:biomed:role=member#aai.egi.eu
    }
    galaxy_admin
    {
      AttributeScope = compute.read compute.write ...
    }  
  }
}
Resources
 {
    IdProviders
    {
        CheckIn
        {
            issuer = ...
            ...
            Clients
            {
                Biomed_User
                {
                    ClientID = ...
                    ClientSecret = ...
                    TokenProfile = AARC
                    VO = biomed 
                }
                Biomed_Admin
               {
                   ClientID = ...
                   ClientSecret = ...
                   # the token profile can be defined at the level of the client, I don't think it would have any sense
                   # but it is technically possible
                   TokenProfile = WLCG
                   VO = biomed
               }
            } 
        } 
        WLCG_IAM
        {
            issuer = ...
            Clients
            {
                DTeam
                {
                    ...
                }
            } 
        }
    }
  }
```

Should this be included in this PR?
May be we should open an issue to further discuss about this?

Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: To summarize the outcome of our discussion with Alexandre on zoom, the following example of the configuration captures  the agreed schema:

```yaml
Registry
{
  VO
  {
    biomed
    {
      IdProvider = CheckIn
      IdPClient = Biomed
    }
    dteam
    {
      IdProvider = WLCG_IAM
      IdPClient = Dteam
    }  
  }
  Groups
  {
    biomed_user
    {
      # The group uses Biomed IdPClient as defined in the biomed VO section 
      # this would be explicitly mapped by OAuth2IdProvider.getUserGroups(in parseIdPMapping),
      # no coded rules or regexps are used
      IdPMapping = eduperson_entitlement:urn:mace:egi.eu:group:registry:biomed:role=member#aai.egi.eu
      # No special other scopes for this group (so far)
    }
    biomed_pilot
    {
       IdPMapping = eduperson_entitlement:urn:mace:egi.eu:group:registry:biomed:role=pilot#aai.egi.eu
       Scope = compute.read compute.write ...
       # By default groups are using the IdProvider defined for the VO. In some cases a different client of the same
       # identity provider can be used
       IdPClient = BiomedPilot 
    }  
    dteam_user
    {
       IdPMapping = wlcg.groups:/dteam/user
    }
  }
}
Resources
 {
    IdProviders
    {
        CheckIn
        {
            issuer = ...
            Grant = device, refresh_token, authorization_code 
            ...
            Clients
            {
                Biomed
                {
                    ClientID = ...
                    ClientSecret = ...
                }
                BiomedPilot
               {
                   ClientID = ...
                   ClientSecret = ...
                   # overriding grant definition of the IdProvider
                   Grant = client_credentials
               }
            } 
        } 
        WLCG_IAM
        {
            issuer = ...
            Clients
            {
                DTeam
                {
                    ...
                }
            } 
        }
    }
  }
```

There are a couple of additions: 1. possibility to define IdProvider Client per group, group can not define another IdProvider than its VO; 2. Grant type that a given Client uses - this is to specify that pilot tokens are obtained with client_credentials, for example.

Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: 2. So IIUC, if one grant is defined, then we use it to get a token, right? If multiple ones are defined, what do we do? Is it only used to check that a requested grant type is allowed?

There is one missing point in your example: how to extract the group name from the right claim (`WLCG`: `wlcg.groups`).
We conclude that we should have a `Claim` or `GroupClaim` option defined in an `IdProvider` section (which could be overridden in a `Client` subsection)

Here is an additional example to take into account: 
I read the Indigo IAM documentation about token profile again, and it seems that a same OAuth2 client might distribute different tokens according to the group-based scope provided.
What if multiple small VOs use the same client but wants to use different token profiles?
Should the `GroupClaim` option be defined in `Resources/IdProviders/<IdProvider>/Clients/<Client>` or in `Registry/VO/<VO>`?

Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: For the grant preference I think we use client_credentials if it is specified in the configuration of the client, otherwise "normal" flows as it is now. 
As for the claim resolution - isn't it enough to take the claim name as the part of IdPMapping before the first colon ? Then this answers also the last question

Comment (atsareg) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: In case of CheckIn the group claim is in the scope actually as for example eduperson_entitlement:urn:mace:egi.eu:group:training.egi.eu:role=member#aai.egi.eu

I guess WLCG IAM does the same ? So then we can look for the IdPMapping in the scopes simply without even extracting the group claim name

Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/IAMIdProvider.py:27]: > For the grant preference I think we use client_credentials if it is specified in the configuration of the client, otherwise "normal" flows as it is now.

So this could be a way to know whether a certain VO would like to use the `client_credentials` flow to submit pilots for instance, couldn't it?
But is there any advantage to set the other flows in this `Grant` option?
One managing the OAuth2 client has to think about updating the DIRAC configuration if there are changes.

> As for the claim resolution - isn't it enough to take the claim name as the part of IdPMapping before the first colon ? Then this answers also the last question

I was thinking about that yes. For the moment, I think it should be enough.

File: src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py (hunk 7 above)
Comment (aldbr) [reply src/DIRAC/Resources/IdProvider/OAuth2IdProvider.py:466]: `OAuthError` is the expected exception that we can get from `authlib` here.
But I don't handle the `requests` exceptions that can also occur while interacting with the IdProvider (like I do in `_handleRequest()`.
I added back a general `except Exception` for the moment.

Not really related to this but:
I wonder whether we really need `OAuth2IdProvider` to inherit from `authlib.OAuth2Session`.
I have the feeling that we could just use the `requests` library to interact with the IdProvider (the code I tested does not absolutely require `authlib`, but I haven't deeply checked the code related to the `device_code` and `authorization_code` flows yet).

**Comment by aldbr** [comment]
One of the integration tests had to be rerun 3 times before passing: it seems flaky.
The error occurred when preparing the environment. The program tries to get an admin token from IAM and get: `requests.exceptions.ConnectionError: ('Connection aborted.', RemoteDisconnected('Remote end closed connection without response'))`
I was unable to reproduce the issue locally.
My guess is that, sometimes, the IAM instance is not fully ready yet, while we try to interact with it.

I added a new commit to fix this problem (based on what was done for the MySQL instance): the first interaction with IAM is tried 10 times. Each time the connection fail, there is a sleep of 10 sec. As soon as the request works, we exit the loop.

**Comment by aldbr** [comment]
@atsareg and I discussed about what should be included in this PR. 
We think that the PR is already large enough and that it should mainly contain the IAM integration tests and minimal changes of the existing code.
The discussion about the structure of the configuration and some parts of the code should come in a different PR.
Any opinion about that?

**Review by fstagni — APPROVED** [review]
